### ISSUE TITLE
With Gamepad plugged in, ET:L does not start - black screen, not responding - SDL BUG #2833

### ISSUE THREAD

**Issue by ETLegacyTeam** [etlegacy/etlegacy#757]
Author: @Kissaki (Kissaki)
Date: 2015-01-05
Redmine Issue: 757, https://dev.etlegacy.com/issues/757

---

With ET:L 2.71a, having my Razer Onza (XBox-style gamepad controller) plugged in ET:L does not start at all.

SDL2.dll seems to continually throw access violation exceptions.

In order to get a stacktrace I built 8ed276 myself with nmake VS2013, built SDL2 from the sources in the lib folder, and replaced the contents of SDL2-windows with the self-compiled SDL2 include files and library files.
The stack trace is attached as a screenshot.

---

Related commit(s):
2e8b4140071dbc8aeaf7a862f5409164df4d6752 - sdl: disable joystick code, refs 757
 cecfa7a6bb10107c72f20b29a325af9561e7d873 - sdl: commented out disabled code, refs 757
 4cd54dc790b0e1a583d9d57d23ffe798be2513c6 - sdl: joystick code enabled again - we no longer in..
 d66999ed98751e1c47f223b2dabd244b5ad95b40 - sdl: completely reenabled joystick code, refs 243..
 b3273168fd1c5cd383b7351c05fd673af63b5ca5 - sdl: don't process the joystick when unfocussed et..

**Comment by ETLegacyTeam** [comment]
From: @Kissaki (Kissaki)
Date: 2015-01-05T16:55:58Z

---

I debugged a bit.

1.)
etl.exe IN_InitJoystick
calls SDL2.dll SDL_Init
calls SDL2.dll SDL_SYS_JoystickInit
calls SDL2.dll SDL_SYS_JoystickDetect
calls IDirectInput8_EnumDevices with a non-NULL dinput

2.)
Through etl.exe IN_InitJoystick
calls SDL2.dll SDL_JoystickQuit
calls SDL2.dll SDL_SYS_JoystickQuit
https://github.com/etlegacy/etlegacy/blob/fb46b5b97201749f792461c63c35484ff1981c60/src/sdl/sdl_input.c#L710-L715

SDL_SYS_JoystickQuit unsets dinput via `IDirectInput8_Release(dinput);`

3.)
Then, during event polling in the main event loop,
etl.exe IN_ProcessEvents 1030
calls SDL2.dll SDL_SYS_JoystickDetect again,
calls IDirectInput8_EnumDevices with with dinput = NULL (after it was released above

Thus, I suspect `IDirectInput8_EnumDevices` dereferences dinput which is NULL, which results in an access violation.
My guess now is that my gamepad is recognized my SDL2, an added event is added to the event queue, but the ETL code releases the joystick before this event is handled.

So I guess, either the early release is not allowed by the SDL, or the SDL is not robust enough.

**Comment by ETLegacyTeam** [comment]
From: @Kissaki (Kissaki)
Date: 2015-01-05T16:58:50Z

---

Attached stacktraces as screenshots.

**Comment by ETLegacyTeam** [comment]
From: @saukko28 (Saukko)
Date: 2015-01-05T17:39:39Z

---

I have a couple of gamepads at home. I had my Logitech Extreme Wingman plugged in for ages without any problems occuring. It is a pretty old gamepad but I have also another one to test with. I will tell tomorrow about the results when I get back home.

-*S

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-01-05T21:49:17Z

---

From IRC:

```
18:01 < Kissaki> currently forumlating a robustness question on their irc :)
18:04 < Ensiform> what SDL2
18:11 < Ensiform> where is LegacySDL_Init
18:12 < Ensiform> according to docs the added NOPARACHUTE is irrelevant :d
18:14 < Ensiform> SDL_Quit is also ignored on windows
18:14 < Kissaki> it's an etl implementation error. SDL guys say only one SDL_Init should be called, and subsystems be initialized 
                 by SDL_InitSubsystem
18:15 < Ensiform> also an ioq3 error

```

SDL "NOPARACHUTE":https://wiki.libsdl.org/MigrationGuide is now build in, and "SDL_InitSubSystem":https://wiki.libsdl.org/SDL_InitSubSystem should be used instead of multiple SDL_Init calls.

**Comment by ETLegacyTeam** [comment]
From: @Kissaki (Kissaki)
Date: 2015-01-05T22:18:27Z

---

Spyhawk wrote:
> 18:14 < Kissaki> it's an etl implementation error. SDL guys say only one SDL_Init should be called, and subsystems be initialized by SDL_InitSubsystem

That’s obsolete.
Initial misinformation by one of the people in their IRC. Was resolved through discussion.

I created a minimal program for reproducibility of the issue and will submit it to the SDL bug tracker.
Before that, will try with their current HEAD version.

As for the not directly associated LegacySDL_Init, I did local commits cleaning that up as well as the incomplete checks of the return code and will try to send that as a PR.

**Comment by ETLegacyTeam** [comment]
From: @Kissaki (Kissaki)
Date: 2015-01-05T22:20:04Z

---

To be more precise, SDL_InitSubsystem should be used, but is not really a requirement (SDL_Init calls that directly anyway).
For clean code, it should - according to the #SDL chatters - be used though.

**Comment by ETLegacyTeam** [comment]
From: @Kissaki (Kissaki)
Date: 2015-01-05T23:09:38Z

---

I created an "SDL bug 2833()":https://bugzilla.libsdl.org/show_bug.cgi?id=2833
with a minimal program that leads to the access violation.

**Comment by ETLegacyTeam** [comment]
From: @Kissaki (Kissaki)
Date: 2015-01-05T23:22:59Z

---

I also created the mentioned pull request for the changes that don’t belong to this issue (just related, and reasonable changes).
https://github.com/etlegacy/etlegacy/pull/110

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2015-01-06T13:31:19Z

---

Since we've had at least 5 reports of crash-at-startup on Windows in the past few days, I disabled the Joystick code for now.
According to SDL #2833 and the attached test program, it seems to really be a SDL2 bug - nothing we can do here.

But maybe it'd be worth adding some macro to enable it for Pandora users only?

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2016-01-04T17:49:07Z

---

From the SDL2 2.04 changelog: Mac OS X:
* Improved joystick hot-plug detection

I don't know if SDL 2.04 fixes the above case Kissaki discovered or the issues with OSX. However our joystick code is inactive (see IN_InitJoystick()) and we should enable it again.
IN_InitJoystick() needs some rework: Regardless how in_joystick cvar is set the code tries to init system joysticks/gamepads. We should check in_joystick at start of IN_InitJoystick(), so there is no system init process when in_joystick is set to 0.

**Comment by ETLegacyTeam** [comment]
From: @IR4T4 (IR4T4)
Date: 2016-01-07T19:37:18Z

---

Joystick code has been enabled again. In case of crash restart ETL:

```
etl[exe] +set in_joystick 0

```

Waiting for the SDL fix - moved to milestone ALL.

**Comment by ETLegacyTeam** [comment]
From: @saukko28 (Saukko)
Date: 2016-11-01T13:00:10Z

---

Could it be that this issue is fixed? At least according to the link below.
https://hg.libsdl.org/SDL/rev/aa816d06ed78

-*S

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2016-11-01T17:11:35Z

---

It will be fixed once a new SDL version is released, and once we do a new release using it.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2016-12-02T18:23:57Z

---

This is fixed in SDL 2.0.5 which has been released a few weeks ago. We should upgrade our SDL for the next version.
Also remove related comment in sdl_input.c.
